### PR TITLE
JMX footprint work

### LINKED ISSUE #133: TableMetricsObjectname is not serializable
It says it is, but is not. APIMBeanServer replaces it in RPC calls, but it is still occasionally seen in Notifications, which causes exceptions in the JMX connector.

### LINKED ISSUE #134: Column family beans refresh is not syncronized
Later JDKs/connections to mbean server can cause multiple threads to call the registration check code simultaneously. While this is not fatal at all, it will cause more object storms and occasionally higher object pressure than needed. 

Should use a lightweight sync and try to avoid unnecessary repeated calls to the API server.

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #133
Fixes #134 
Refs #135

Makes CF mbean refresh code synchronized and tries to remove reductant calls if we contend. 
Adds background reaping of dead objects to reduce memory load in (test) scenarios where we manage
to refresh to add, but not cause removal (i.e. no wildcard queries). 

TableMetricsObjectName serialization is fixed in the series because without it we see loads of 
exceptions when refreshing the mbean set.

### CODE REVIEW

File: src/main/java/org/apache/cassandra/metrics/TableMetrics.java [hunk 1]
```diff
@@ -400,6 +401,18 @@ public boolean isPropertyListPattern() {
...
+         * This type is not really serializable. 
+         * Replace it with vanilla objectname.
+         */
+		private Object writeReplace() throws ObjectStreamException {
+			try {
+				return new ObjectName(getDomain(), getKeyPropertyList());
+			} catch (MalformedObjectNameException e) {
+				throw new InvalidObjectException(toString());
+			}
+		}
```
Comment (penberg) [inline src/main/java/org/apache/cassandra/metrics/TableMetrics.java:415]: Did you discover the one true indentation style -- tabs?

File: src/main/java/com/scylladb/jmx/metrics/RegistrationChecker.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		check(client, server, REMOVE_NO_WAIT);
+	}
+
+	public final void check(APIClient client, JmxMBeanServer server) throws OperationsException, UnknownHostException {
+		check(client, server, ADD_AND_REMOVE);
+	}
+
+	public final void check(APIClient client, JmxMBeanServer server, EnumSet<RegistrationMode> mode)
+			throws OperationsException, UnknownHostException {
+		if (!lock.tryLock() && mode.contains(Wait)) {
```
Comment (penberg) [inline src/main/java/com/scylladb/jmx/metrics/RegistrationChecker.java:35]: If `tryLock` fails, but we're not allowed to wait, what is supposed to happen? Right now, we perform `doCheck` without the lock (which I assume is safe?), but we also call `unlock()`...

**Review by penberg — COMMENTED** [review]
Can you also say few words on why we need `RegistrationChecker`? Why do we need light-weight synchronization for this, instead of slapping `synchronized` everywhere?

File: src/main/java/org/apache/cassandra/metrics/TableMetrics.java (hunk 1 above)
Comment (elcallio) [reply src/main/java/org/apache/cassandra/metrics/TableMetrics.java:415]: Meh. Ran a new eclipse. We should check the formatter in with the maven project.

**Comment by elcallio** [comment]
Fixed the above notes.

**Pull request merged**